string msg
---
string ret_msg